src/parser/mod.rs

```diff
@@ -0,0 +1,695 @@
+			DefaultTranslator.translate_code(funcion_body, &mut code_builder)?
+		}
+
+		for (func_type, func_body) in funcs {
+			let func_type_index = self.add_func_type(func_type)?;
+
+			// Define a new function in Function section
+			function_builder.function(func_type_index);
+			self.function_map.push(func_type_index);
+
+			// Write new function body in Code section
+			code_builder.function(func_body);
+			self.func_bodies_count += 1;
+		}
+		self.replace_section(SectionId::Function.into(), &function_builder)?;
+		self.replace_section(SectionId::Code.into(), &code_builder)
+	}
+
+	pub fn add_import_func(
+		&mut self,
+		module: &str,
+		func_name: &str,
+		func_type: Type,
+	) -> Result<()> {
+		let func_type_idx = self.add_func_type(&func_type)?;
+
+		// Recreate Import section
+		let mut import_decoder = wasm_encoder::ImportSection::new();
+		for import in self.import_section()? {
+			DefaultTranslator.translate_import(import, &mut import_decoder)?;
+		}
+
+		// Define new function import in the Import section.
+		import_decoder.import(module, func_name, wasm_encoder::EntityType::Function(func_type_idx));
+		// function_map consist of:
+		// - imported function first
+		// - local functions then
+		// This is important when getting function Type by its index get_type_by_func_idx()
+		// When adding an import we make sure it is added as the last imported function
+		// but before local functions
+		self.function_map.insert(self.imported_functions_count as usize, func_type_idx);
+
+		self.imported_functions_count += 1;
+		self.replace_section(SectionId::Import.into(), &import_decoder)
+	}
+
+	pub fn add_import_global(
+		&mut self,
+		module: &str,
+		global_name: &str,
+		global_type: GlobalType,
+	) -> Result<()> {
+		// Recreate Import section
+		let mut import_decoder = wasm_encoder::ImportSection::new();
+		for import in self.import_section()? {
+			DefaultTranslator.translate_import(import, &mut import_decoder)?;
+		}
+
+		// Define new global import in the Import section.
+		import_decoder.import(
+			module,
+			global_name,
+			DefaultTranslator.translate_global_type(&global_type)?,
+		);
+		self.imported_globals_count += 1;
+		self.replace_section(SectionId::Import.into(), &import_decoder)
+	}
+
+	pub fn modify_memory_type(&mut self, mem_index: u32, mem_type: MemoryType) -> Result<()> {
+		let mut memory_builder = wasm_encoder::MemorySection::new();
+		let mut memory_types = vec![];
+		for (index, memory) in self.memory_section()?.iter().enumerate() {
+			let encoded_mem_type = if index as u32 != mem_index {
+				memory_types.push(*memory);
+				DefaultTranslator.translate_memory_type(memory)?
+			} else {
+				memory_types.push(mem_type);
+				DefaultTranslator.translate_memory_type(&mem_type)?
+			};
+			memory_builder.memory(encoded_mem_type);
+		}
+		self.memory_types = memory_types;
+		self.replace_section(SectionId::Memory.into(), &memory_builder)
+	}
+
+	pub fn bytes(&self) -> Vec<u8> {
```

Comment:
Why didn't you follow the original implementation by serialising module in a defined order?

The original code has a note: `// datacount goes before code`. Do we understand why this?

Reply:
Good catch! It is clearly a mistake.

When working with `fvm-wasm-instrument` repo I was moving between commits. And I accidentally copied this function from different revision.
This is the [commit](https://github.com/filecoin-project/fvm-wasm-instrument/pull/11/commits/f9fa538f72db3c581fb0c6747fce18192c59a3af) where it was fixed 
I will fix it as well.

`// datacount goes before code` -  According to the WASM spec the data count section occurs before the code section, so a single-pass validator can use this count instead of deferring validation.
see [Note! section](https://webassembly.github.io/spec/core/binary/modules.html#data-count-section)

Reply:
Done